src/integrations/cleanup-integration.php

```diff
@@ -25,63 +32,236 @@ public static function get_conditionals() {
 	 * @return void
 	 */
 	public function register_hooks() {
-		\add_action( 'wpseo_cleanup_indexables', [ $this, 'cleanup_obsolete_indexables' ], 10, 2 );
-		\add_action( 'wpseo_cleanup_orphaned_indexables', [ $this, 'cleanup_orphaned_indexables' ] );
-		\add_action( 'wpseo_deactivate', [ $this, 'unschedule_cron' ] );
+		\add_action( self::START_HOOK, [ $this, 'run_cleanup' ] );
+		\add_action( self::CRON_HOOK, [ $this, 'run_cleanup_cron' ] );
+		\add_action( 'wpseo_deactivate', [ $this, 'reset_cleanup' ] );
 	}
 
 	/**
-	 * Cleans rows from the indexable table and unregisters the cron if no deletions.
+	 * Returns the conditionals based on which this loadable should be active.
 	 *
-	 * @param string $object_type     The object type to query.
-	 * @param string $object_sub_type The object subtype to query.
+	 * @return array The array of conditionals.
+	 */
+	public static function get_conditionals() {
+		return [];
+	}
+
+	/**
+	 * Starts the indexables cleanup.
 	 *
 	 * @return void
 	 */
-	public function cleanup_obsolete_indexables( $object_type, $object_sub_type ) {
-		$number_of_deletions = $this->clean_indexables_with_object_type( $object_type, $object_sub_type, 1000 );
+	public function run_cleanup() {
+		$this->reset_cleanup();
+
+		$cleanups = $this->get_cleanup_tasks();
+		$limit    = $this->get_limit();
+
+		foreach ( $cleanups as $name => $action ) {
+			$items_cleaned = $action( $limit );
+
+			if ( $items_cleaned === false ) {
+				return;
+			}
 
-		if ( empty( $number_of_deletions ) ) {
-			$this->unschedule_cron( 'wpseo_cleanup_indexables' );
+			if ( $items_cleaned < $limit ) {
+				continue;
+			}
+
+			// There are more items to delete for the current cleanup job, start a cronjob at the specified job.
+			$this->start_cron_job( $name );
+			return;
 		}
 	}
 
 	/**
-	 * Cleans orphaned rows from the yoast tables and unregisters the cron if no deletions.
+	 * Returns an array of cleanup tasks.
+	 *
+	 * @return Closure[] The cleanup tasks.
+	 */
+	protected function get_cleanup_tasks() {
+		return \array_merge(
+			[
+				'clean_indexables_with_object_type_and_object_sub_type_shop_order' => function( $limit ) {
+					return $this->clean_indexables_with_object_type_and_object_sub_type( 'post', 'shop_order', $limit );
+				},
+				'clean_indexables_by_post_status_auto-draft' => function( $limit ) {
+					return $this->clean_indexables_with_post_status( 'auto-draft', $limit );
+				},
+			],
+			/**
+			 * Filter: Adds the possibility to add addition cleanup functions.
+			 *
+			 * @api array Associative array with unique keys. Value should be a cleanup function that receives a limit.
+			 */
+			\apply_filters( 'wpseo_cleanup_tasks', [] ),
```

Comment:
I would test whether the result of applying the filter is valid before using it. We do not know who is going to hook into the filter so it is safer to err on the side of caution.

E.g. a sanity check if it is an array, if not: default to an empty array.